### ISSUE TITLE
Printout, property chains, and auxiliary filters

### ISSUE THREAD

**Issue by mwjames** [SemanticMediaWiki/SemanticMediaWiki#2655]
Relates to #1824 and https://sourceforge.net/p/semediawiki/mailman/message/36013326/ with

> Is it possible to specify which object to access if I
have multiple subobjects with different names (e.g., SubExample1 and
SubExample2)?

```
[[Name::Example A]]
[[Date::1/1/2018]]

{{#subobject:SubExample1
 |count=20
 |value=40}}{{#subobject:SubExample2
 |count=20
 |value=41}}{{#subobject:SubExample3
 |count=25
 |value=42
}}
```

## Notes

A general note, using property chains in a printout (#1824) represents a post-processing traversal on the result set that was matched to the initial query condition therefore trying the reduce the set of value entities during the post process is not possible unless an additional syntax is introduced to allow to specify some auxiliary conditions.

One such syntax is being proposed in following example.

```
{{#ask: [[Name::Example A]]
 |?Date
 |?Has subobject.Count|+filter=SubExample2,SubExample1
 |?Has subobject.Value|+filter=SubExample3,SubExample2,SubExample1
}}
```

## Example code

The following changes have not been tested and currently there is no definition as to what `|+filter=...` is expected to do or how it should be validated.

The prototype implementation assumes a comma separated list that is matched against the subobject name when `Has subobject` is used as chain element and excludes any matched entities from the value result set and hereby any traversal function of succeeding chain members.

```
@@ -203,36 +203,38 @@ class ResultFieldMatchFinder {
 			$propertyValue = $this->printRequest->getData();
 		}
 
 		$index = $this->printRequest->getParameter( 'index' );
 		$lang = $this->printRequest->getParameter( 'lang' );
+
 		$newcontent = array();
+		$property = $propertyValue->getDataItem();
 
 		// Replace content with specific content from a Container/MultiValue
 		foreach ( $content as $diContainer ) {
 
 			/* AbstractMultiValue */
 			$multiValue = DataValueFactory::getInstance()->newDataValueByItem(
 				$diContainer,
-				$propertyValue->getDataItem()
+				$property
 			);
 
 			$multiValue->setOption( $multiValue::OPT_QUERY_CONTEXT, true );
 
 			if ( $multiValue instanceof MonolingualTextValue && $lang !== false && ( $textValue = $multiValue->getTextValueByLanguage( $lang ) ) !== null ) {
 
 				// Return the text representation without a language reference
 				// (tag) since the value has been filtered hence only matches
 				// that language
-				$newcontent[] = $this->applyContentManipulation( $textValue->getDataItem() );
+				$newcontent[] = $this->applyContentManipulation( $property, $textValue->getDataItem() );
 
 				// Set the index so ResultArray::getNextDataValue can
 				// find the correct PropertyDataItem (_TEXT;_LCODE) position
 				// to match the DI
 				$this->printRequest->setParameter( 'index', 1 );
 			} elseif ( $lang === false && $index !== false && ( $dataItemByRecord = $multiValue->getDataItemByIndex( $index ) ) !== null ) {
-				$newcontent[] = $this->applyContentManipulation( $dataItemByRecord );
+				$newcontent[] = $this->applyContentManipulation( $property, $dataItemByRecord );
 			}
 		}
 
 		$content = $newcontent;
 		unset( $newcontent );
@@ -281,20 +283,22 @@ class ResultFieldMatchFinder {
 	}
 
 	private function doFetchPropertyValues( $dataItems, $dataValue ) {
 
 		$propertyValues = array();
+		$property = $dataValue->getDataItem();
+		$isChain = $this->printRequest->isMode( PrintRequest::PRINT_CHAIN );
 
 		foreach ( $dataItems as $dataItem ) {
 
 			if ( !$dataItem instanceof DIWikiPage ) {
 				continue;
 			}
 
 			$pv = $this->store->getPropertyValues(
 				$dataItem,
-				$dataValue->getDataItem(),
+				$property,
 				$this->getRequestOptions()
 			);
 
 			if ( $pv instanceof \Iterator ) {
 				$pv = iterator_to_array( $pv );
@@ -302,18 +306,32 @@ class ResultFieldMatchFinder {
 
 			$propertyValues = array_merge( $propertyValues, $pv );
 			unset( $pv );
 		}
 
-		array_walk( $propertyValues, function( &$dataItem ) {
-			$dataItem = $this->applyContentManipulation( $dataItem );
-		} );
+		foreach ( $propertyValues as $key => $value ) {
+			if ( ( $dataItem = $this->applyContentManipulation( $property, $value, $isChain ) ) !== null ) {
+				$propertyValues[$key] = $dataItem;
+			} else {
+				unset( $propertyValues[$key] );
+			}
+		}
 
 		return $propertyValues;
 	}
 
-	private function applyContentManipulation( $dataItem ) {
+	private function applyContentManipulation( $property, $dataItem, $isChain = false ) {
+
+		if ( $isChain && ( $filter = $this->printRequest->getParameter( 'filter' ) ) !== false ) {
+
+			// Apply filter on something like ?Has subobject.Foo|+filter=Foobar,Bar
+			$excludeList = explode( ',', $filter );
+
+			if ( $property->getKey() === '_SOBJ' && in_array( $dataItem->getSubobjectName(), $excludeList ) ) {
+				return null;
+			}
+		}
 
 		if ( !$dataItem instanceof DIBlob ) {
 			return $dataItem;
 		}
 ```

## Questions

* Does it make sense to support such use case?
* What is expected of something like ` ?Has subobject.Foo|+filter=Foobar,Bar`?
* Are there any other uses cases that would benefit from an auxiliary filter on a property path?

**Comment by jaideraf** [comment]
@mwjames Wouldn't `|+filter=` collide with Filtered format parameters? [0]
[0] https://www.semantic-mediawiki.org/wiki/Help:Filtered_format

**Comment by kghbln** [comment]
I believe `|+filter=` will collide but this is just the wording. We could use "use", "select", "pick", "take", etc.

What comes to my mind as a possible use case is data evolution over time, i.e. every named subobject represents a point in time. So I will be able to e.g. add currrent addresses for a person and still preserve the old ones and at the same time query for the latest subobject. I think this is not the best of examples but ...

**Comment by jaideraf** [comment]
@kghbln Interesting example. Maybe not related to this proposal, but somehow it reminds me the `Rank` feature of Wikibase. [1]
The address could be "_preferred_", "_normal_" or "_deprecated_".

[1] https://www.wikidata.org/wiki/Wikidata:Glossary#Rank

**Comment by mwjames** [comment]
To avoid this being lost, I added #2667 as example implementation which also incorporates changes from below.

> Wouldn't |+filter= collide with Filtered format parameters?

Now that you mentioned it, using a different identifier shouldn't be an issue but I'm thinking to enforce some sort of naming convention to avoid trouble in future where other parameters are introduced by for example SRF.

Something like `|+Has subobject.filter= ...` where `property label` + `.` + `filter` are concatenated and used as identifying element for a path filter to avoid any ambiguity on the selected instance as demonstrated by:

```
{{#ask: [[Name::Example A]]
 |?Date
 |?Has subobject.Count|+Has subobject.filter=SubExample2,SubExample1
 |?Has subobject.Value|+Has subobject.filter=SubExample3,SubExample2,SubExample1
}}
```

Above translates into "show the count value of the selected subobjects but only show counts on subobjects that match the internal name of `SubExample2` and `SubExample1`". You can also do a simple value match such as "show only those that match 40 for the value property".

```
{{#ask: [[Name::Example A]]
 |?Date
 |?Has subobject.Value|+Value.filter=40
}}
```

Things like `|+Has subobject.Value.filter` are not valid with a filter expecting only to use the last matchable path member.

**Comment by schordave** [comment]
> Does it make sense to support such use case?
> Are there any other uses cases that would benefit from an auxiliary filter on a property path?

I was the original OP that asked about it initially. Such mechanism is necessary. I'll give you a practical example that we have been trying to solve.

Consider the following page [0]. It's got many specifications which are grouped into subojects (although we have not grouped everything yet because this feature does not exist so it's unclear what direction we're trying to go). One common suboject found on roughly 99% of our similar articles is the package object [1].

A certain model could actually have multiple packages and each once with entirely different specs which means referring unambiguously to a specific package (either by order, e.g., "0", or by some name, e.g., "package 0") is a necessity for us. Consider this more complex article [2] where we, in theory, we would have multiple memory controllers subobjects and multiple GPUs subobjects. In this case, it makes no sense to be able to even query all subojects at once because the request is meaningless without knowing that we're querying the same specific suboject.

Basically, without the ability to filter by specific subojects such queries are practically broken for us. Unless of course, you think there is something else we can use to achieve what we are after.

[0] https://en.wikichip.org/wiki/intel/core_i9/i9-8950hk
[1] https://en.wikichip.org/wiki/Special:Browse/:intel-2Fcore-20i9-2Fi9-2D8950hk-23package
[2] https://en.wikichip.org/wiki/intel/core_i7/i7-8705g

**Comment by mwjames** [comment]
Moving it to "Enhancements and features" for future discussions as it unlikely to be part of the 3.0 release.
